Add ADC Dual Mode support with new AdvancedADCDual class

jmdodd95682 opened the pull request:
The proposed change adds a new class `AdvancedADCDual` class which is used to handle configuring ADC1 and ADC2 and starting them synchronously by enabling Dual mode in the ADC1/2_CCR register.

It works like this:

```cpp
AdvancedADC adc_input[2];
AdvancedADCDual adc_dual;

result = adc_dual.begin(&(adc_input[0]), &(adc_input[1]), sample_resolution, sample_rate, num_samples, QUEUE_DEPTH, 2, &pins);
```

where `pins` is an array of pins numbers.

The AdvancedADCDual `begin()` routine calls the AdvancedADC `begin()` for each instance.  
It configures the first pin in the array on ADC1 and the remaining pins on ADC2.  There was not a cleaner way to do this
without passing two arrays (which is possible).  Finally it calls `start()` on the AdvancedADC mapped to ADC1.

The `AdvancedADC.begin()` routine was modified to take two optional parameters: `do_start` and `adcNum`.
Additionally, a new `start()` routine was added to `AdvancedADC()`.
The `do_start` parameter allows the AdvancedADCDual to hold-off starting the ADC until both are configured.
The `adcNum` is required because you must force the input pins to be mapped to ADC1 and ADC2 for dual mode.
Since both parameters are defaulted, the `AdvancedADC.begin()` remains backward compatible with existing code.

Finally, the `HALConfig.cpp` and `.h` were modified to add two new routines: `hal_enable_dual_mode` and `hal_disable_dual_mode`.
These access the low-level driver to write a 5-bit value to the ADC1/2_CCR register enabling or disabling Dual Mode Simultaneous.

I've tested this, and it seems to work properly.

iabdalkader commented:
Hi! Thanks for sending this! I'll test it tomorrow. In the meantime, do we need to specify the ADC number ? See my comment here: https://github.com/arduino-libraries/Arduino_AdvancedAnalog/pull/60#issuecomment-2004152259

Also, can you please format your code like the rest of the library ? I'll leave a few hints in a review.

iabdalkader commented on src/AdvancedADC.cpp:
```diff
@@ -115,42 +115,73 @@ DMABuffer<Sample> &AdvancedADC::read() {
 }
 
-int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers) {
+int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers,bool do_start,uint8_t adcNum) {
```
```suggestion
int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers,bool do_start, uint8_t adc_num) {
```
Space between args, and No camel case please.

iabdalkader commented on src/AdvancedADC.cpp:
```diff
@@ -115,42 +115,73 @@ DMABuffer<Sample> &AdvancedADC::read() {
     }
 
+    //if ADC specified is more than the number of available ADC bail out
+    if(adcNum>AN_ARRAY_SIZE(adc_pin_alt))
+    {
+        descr = nullptr;
+        return 0;
+    }
```
```suggestion
    if (adcNum > AN_ARRAY_SIZE(adc_pin_alt)) {
        descr = nullptr;
        return 0;
    }
```

Spaces between operators, bracket on the same line as function/if.

iabdalkader commented on src/AdvancedADC.cpp:
```diff
@@ -115,42 +115,73 @@ DMABuffer<Sample> &AdvancedADC::read() {
             }
         }
     }
+    else if(adcNum>0)  //if ADC specified use that ADC to try to map first channel
```
```suggestion
    } else if (adcNum > 0) {  //if ADC specified use that ADC to try to map first channel
```

iabdalkader commented on src/AdvancedADC.cpp:
```diff
@@ -115,42 +115,73 @@ DMABuffer<Sample> &AdvancedADC::read() {
+                adc_pins[0] = pin;
+            }
+        }
+        selectedADC=adcNum; //Store selected number
```
```suggestion
        selected_adc = adc_num; //Store selected number
```

iabdalkader commented on src/AdvancedADC.cpp:
```diff
@@ -212,12 +273,24 @@ int AdvancedADC::begin(uint32_t resolution, uint32_t sample_rate, size_t n_sampl
     hal_dma_enable_dbm(&descr->dma, descr->dmabuf[0]->data(), descr->dmabuf[1]->data());
     HAL_NVIC_EnableIRQ(descr->dma_irqn);
 
+    if(do_start)
+    {
+        return(start(sample_rate));
+    }
+
```
```suggestion
    if (do_start) {
        return start(sample_rate);
    }
```

No `return()` please.

iabdalkader commented on src/HALConfig.cpp:
```diff
@@ -167,10 +167,22 @@ static uint32_t ADC_RANK_LUT[] = {
     // Set ADC clock source.
     __HAL_RCC_ADC_CONFIG(RCC_ADCCLKSOURCE_CLKP);
-
+   
```
```suggestion
```
Please remove extra spaces that you added.

iabdalkader commented on src/HALConfig.cpp:
```diff
@@ -206,7 +218,7 @@ int hal_adc_config(ADC_HandleTypeDef *adc, uint32_t resolution, uint32_t trigger
     sConfig.SingleDiff   = ADC_SINGLE_ENDED;
     sConfig.SamplingTime = ADC_SAMPLETIME_8CYCLES_5;
-
+    
```
```suggestion
```

iabdalkader commented on src/HALConfig.cpp:
```diff
@@ -46,18 +46,18 @@ int hal_tim_config(TIM_HandleTypeDef *tim, uint32_t t_freq) {
 
     if (tim->Instance == TIM1) {
-        __HAL_RCC_TIM1_CLK_ENABLE();
+                __HAL_RCC_TIM1_CLK_ENABLE();
```
```suggestion
        __HAL_RCC_TIM1_CLK_ENABLE();
```

Please remove any spaces you added.

iabdalkader commented on src/AdvancedADC.h:
```diff
@@ -41,21 +43,74 @@ class AdvancedADC {
+        void clear();
+
+        int start(uint32_t sample_rate);
+        
+        //void setADCDualMode(bool dm);
+        //int enableDualMode();
+        //int disableDualMode();
+        
```
```suggestion
```

iabdalkader commented on src/AdvancedADC.h:
```diff
@@ -41,21 +43,74 @@ class AdvancedADC {
+		private:
+			size_t n_channels;
+			pin_size_t adc_pins_unmapped[AN_MAX_ADC_CHANNELS];
+			AdvancedADC *adcIN1;
+			AdvancedADC *adcIN2;
```
```suggestion
			AdvancedADC &adc1;
			AdvancedADC &adc2;
```
Those should be passed by reference.

iabdalkader commented on src/AdvancedADC.h:
```diff
@@ -41,21 +43,74 @@ class AdvancedADC {
         SampleBuffer read();
-        int begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers);
-        int begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers, size_t n_pins, pin_size_t *pins) {
+        int begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers, bool do_start=true,uint8_t adcNum=0);
```
```suggestion
        int begin(uint32_t resolution, uint32_t sample_rate, size_t n_samples, size_t n_buffers, bool start=true, uint8_t adc_instance=0);
```

jmdodd95682 commented:
I could not locate your specific comment on the ADC number in #60.  Probably me.

Anyway, the reason we need the ADC number is  that Dual mode only works if you can use ADC1 and ADC2.  It does NOT work with ADC3.  Nor can you combine ADC1 and 3...etc.    

So, if someone has already declared an instance of AdvancedADC and executed a begin() then probably ADC1 will be taken.  And then ADC dual will not work.  The existing begin() will just try to use ADC2 and ADC3, and programming the Dual field will make ADC1 no longer work correctly.  And ADC3 will never start and won't be synchronized.

In addition, the start MUST be done to ADC1 (not ADC2), so the order in which you declare them matters somewhat if your application wants to start capturing based on a triggering condition (e.g. rising edge) of whatever isconnected to ADC1.

So, I THINK you need the adc number specified. for each begin() to make Dual mode work reliably.

iabdalkader commented:
> So, if someone has already declared an instance of AdvancedADC and executed

This would be a user error. We can check if the instances are correct, and check which one is ADC1 and call start on that. And we'll add a comment about that in the example.

Here's my comment from #60 

> Note I don't think we need to specify the ADC number, which should make things easier. The first pin always determines the ADC instance, and the rest of the pins must map to the same ADC (this code is already implemented). The instances/descriptors are also allocated in order, so you'll get ADC1 first, then ADC2, then ADC3. The AdvancedADCDual can also check if they are the correct ADCs or not.

jmdodd95682 commented:
Fair enough.  Yes, you can do it this way.  It does end up in the same place, without the need for another parameter being passed to begin().  Either way if you ask to ADC1 and ADC2 and cannot get it, it just gives you an error.  Good point.

I can make this change.

jmdodd95682 commented:
Okay.  I removed the adc_num from the AdvancedADC.begin().  

The AdvancedADCDual.begin() now checks that the assigned ADC for the first pin is ADC1 and all subsequent pins is ADC2.
The changes to AdvancedADC.begin() are now much simpler.

Also added a new member function in AdvancedADC.getAssignedADC() so that AdvancedADCDual can determine which ADC was assigned and flag an error if its not correct.

I tested it, it works reliably on my app.

iabdalkader commented:
I'll refactor this PR, rebase it and push it again.

per1234 reviewed (commented):
Closing as superseded by https://github.com/arduino-libraries/Arduino_AdvancedAnalog/pull/65

Thanks for your contribution @jmdodd95682!
